// vike/dist/cjs/node/plugin/plugins/importUserCode/v1-design/getVikeConfig/loadFileAtConfigTime.js
"use strict";
// Files loadded at config time:
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.loadImportedFile = loadImportedFile;
exports.loadValueFile = loadValueFile;
exports.loadConfigFile = loadConfigFile;
const utils_js_1 = require("../../../../utils.js");
const transpileAndExecuteFile_js_1 = require("./transpileAndExecuteFile.js");
const assertPlusFileExport_js_1 = require("../../../../../../shared/page-configs/assertPlusFileExport.js");
const picocolors_1 = __importDefault(require("@brillout/picocolors"));
const transformPointerImports_js_1 = require("./transformPointerImports.js");
const getConfigFileExport_js_1 = require("../getConfigFileExport.js");
const resolvePointerImport_js_1 = require("./resolvePointerImport.js");
(0, utils_js_1.assertIsNotProductionRuntime)();
// Load pointer import
async function loadImportedFile(import_, userRootDir, importedFilesLoaded) {
    const f = import_.filePathAbsoluteFilesystem;
    if (!importedFilesLoaded[f]) {
        importedFilesLoaded[f] = (0, transpileAndExecuteFile_js_1.transpileAndExecuteFile)(import_, userRootDir, false).then((r) => r.fileExports);
    }
    const fileExports = await importedFilesLoaded[f];
    const fileExport = fileExports[import_.fileExportName];
    return fileExport;
}
// Load +{configName}.js
async function loadValueFile(interfaceValueFile, configName, userRootDir) {
    const { fileExports } = await (0, transpileAndExecuteFile_js_1.transpileAndExecuteFile)(interfaceValueFile.filePath, userRootDir, false);
    const { filePathToShowToUser } = interfaceValueFile.filePath;
    (0, assertPlusFileExport_js_1.assertPlusFileExport)(fileExports, filePathToShowToUser, configName);
    Object.entries(fileExports).forEach(([exportName, configValue]) => {
        const configName_ = exportName === 'default' ? configName : exportName;
        interfaceValueFile.fileExportsByConfigName[configName_] = { configValue };
    });
}
// Load +config.js, including all its extends pointer imports
async function loadConfigFile(configFilePath, userRootDir, visited, isExtensionConfig) {
    const { filePathAbsoluteFilesystem } = configFilePath;
    assertNoInfiniteLoop(visited, filePathAbsoluteFilesystem);
    const { fileExports } = await (0, transpileAndExecuteFile_js_1.transpileAndExecuteFile)(configFilePath, userRootDir, isExtensionConfig ? 'is-extension-config' : true);
    const { extendsConfigs, extendsFilePaths } = await loadExtendsConfigs(fileExports, configFilePath, userRootDir, [
        ...visited,
        filePathAbsoluteFilesystem
    ]);
    const configFile = {
        fileExports,
        filePath: configFilePath,
        extendsFilePaths
    };
    return { configFile, extendsConfigs };
}
function assertNoInfiniteLoop(visited, filePathAbsoluteFilesystem) {
    const idx = visited.indexOf(filePathAbsoluteFilesystem);
    if (idx === -1)
        return;
    const loop = visited.slice(idx);
    (0, utils_js_1.assert)(loop[0] === filePathAbsoluteFilesystem);
    (0, utils_js_1.assertUsage)(idx === -1, `Infinite extends loop ${[...loop, filePathAbsoluteFilesystem].join('>')}`);
}
async function loadExtendsConfigs(configFileExports, configFilePath, userRootDir, visited) {
    const extendsPointerImportData = getExtendsPointerImportData(configFileExports, configFilePath);
    const extendsConfigFiles = [];
    extendsPointerImportData.map((pointerImportData) => {
        const filePath = (0, resolvePointerImport_js_1.resolvePointerImport)(pointerImportData, configFilePath, userRootDir);
        (0, utils_js_1.assert)(filePath.filePathAbsoluteFilesystem);
        extendsConfigFiles.push(filePath);
    });
    const extendsConfigs = [];
    await Promise.all(extendsConfigFiles.map(async (configFilePath) => {
        const result = await loadConfigFile(configFilePath, userRootDir, visited, true);
        extendsConfigs.push(result.configFile);
        extendsConfigs.push(...result.extendsConfigs);
    }));
    const extendsFilePaths = extendsConfigFiles.map((f) => f.filePathAbsoluteFilesystem);
    return { extendsConfigs, extendsFilePaths };
}
function getExtendsPointerImportData(configFileExports, configFilePath) {
    const { filePathToShowToUser } = configFilePath;
    const configFileExport = (0, getConfigFileExport_js_1.getConfigFileExport)(configFileExports, filePathToShowToUser);
    const wrongUsage = `${filePathToShowToUser} sets the config ${picocolors_1.default.cyan('extends')} to an invalid value, see https://vike.dev/extends`;
    let extendList;
    if (!('extends' in configFileExport)) {
        return [];
    }
    else if ((0, utils_js_1.hasProp)(configFileExport, 'extends', 'string')) {
        extendList = [configFileExport.extends];
    }
    else if ((0, utils_js_1.hasProp)(configFileExport, 'extends', 'string[]')) {
        extendList = configFileExport.extends;
    }
    else {
        (0, utils_js_1.assertUsage)(false, wrongUsage);
    }
    const extendsPointerImportData = extendList.map((importString) => {
        const pointerImportData = (0, transformPointerImports_js_1.parsePointerImportData)(importString);
        (0, utils_js_1.assertUsage)(pointerImportData, wrongUsage);
        return pointerImportData;
    });
    return extendsPointerImportData;
}
