// vike/dist/cjs/node/plugin/plugins/autoFullBuild.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.autoFullBuild = autoFullBuild;
const vite_1 = require("vite");
const utils_js_1 = require("../utils.js");
const runPrerender_js_1 = require("../../prerender/runPrerender.js");
const getConfigVike_js_1 = require("../../shared/getConfigVike.js");
const isViteCliCall_js_1 = require("../shared/isViteCliCall.js");
const picocolors_1 = __importDefault(require("@brillout/picocolors"));
const logErrorHint_js_1 = require("../../runtime/renderPage/logErrorHint.js");
const buildConfig_js_1 = require("./buildConfig.js");
let forceExit = false;
function autoFullBuild() {
    let config;
    let configVike;
    return [
        {
            name: 'vike:autoFullBuild',
            apply: 'build',
            enforce: 'pre',
            async configResolved(config_) {
                configVike = await (0, getConfigVike_js_1.getConfigVike)(config_);
                config = config_;
                abortViteBuildSsr(configVike);
            },
            writeBundle: {
                /* We can't use this because it breaks Vite's logging. TODO: try again with latest Vite version.
                sequential: true,
                order: 'pre',
                */
                async handler(_options, bundle) {
                    try {
                        await triggerFullBuild(config, configVike, bundle);
                    }
                    catch (err) {
                        // Avoid Rollup prefixing the error with [vike:autoFullBuild], for example see https://github.com/vikejs/vike/issues/472#issuecomment-1276274203
                        console.error(err);
                        process.exit(1);
                    }
                }
            }
        },
        {
            name: 'vike:autoFullBuild:forceExit',
            apply: 'build',
            enforce: 'post',
            closeBundle: {
                sequential: true,
                order: 'post',
                handler() {
                    if (forceExit) {
                        (0, runPrerender_js_1.runPrerender_forceExit)();
                    }
                }
            }
        }
    ];
}
async function triggerFullBuild(config, configVike, bundle) {
    if (config.build.ssr)
        return; // already triggered
    if (isDisabled(configVike))
        return;
    // Workaround for @vitejs/plugin-legacy
    //  - The legacy plugin triggers its own Rollup build for the client-side.
    //  - The legacy plugin doesn't generate a manifest => we can use that to detect the legacy plugin build.
    //  - Issue & reproduction: https://github.com/vikejs/vike/issues/1154#issuecomment-1965954636
    if (!bundle[buildConfig_js_1.manifestTempFile])
        return;
    const configFromCli = !(0, isViteCliCall_js_1.isViteCliCall)() ? null : (0, isViteCliCall_js_1.getViteConfigFromCli)();
    const configInline = {
        ...configFromCli,
        configFile: configFromCli?.configFile || config.configFile,
        root: config.root,
        build: {
            ...configFromCli?.build
        }
    };
    try {
        await (0, vite_1.build)({
            ...configInline,
            build: {
                ...configInline.build,
                ssr: true
            }
        });
    }
    catch (err) {
        console.error(err);
        (0, logErrorHint_js_1.logErrorHint)(err);
        process.exit(1);
    }
    if (configVike.prerender && !configVike.prerender.disableAutoRun && configVike.disableAutoFullBuild !== 'prerender') {
        await (0, runPrerender_js_1.runPrerenderFromAutoFullBuild)({ viteConfig: configInline });
        forceExit = true;
    }
}
function abortViteBuildSsr(configVike) {
    if (configVike.disableAutoFullBuild !== true && (0, isViteCliCall_js_1.isViteCliCall)() && (0, isViteCliCall_js_1.getViteConfigFromCli)()?.build.ssr) {
        (0, utils_js_1.assertWarning)(false, `The CLI call ${picocolors_1.default.cyan('$ vite build --ssr')} is superfluous since ${picocolors_1.default.cyan('$ vite build')} also builds the server-side. If you want two separate build steps then use https://vike.dev/disableAutoFullBuild or use Vite's ${picocolors_1.default.cyan('build()')} API.`, { onlyOnce: true });
        process.exit(0);
    }
}
function isDisabled(configVike) {
    const { disableAutoFullBuild } = configVike;
    if (disableAutoFullBuild === null || disableAutoFullBuild === 'prerender') {
        // TODO/v1-release: remove autoFullBuild for both Vite's build() API and Vite's CLI
        //  - Tell users to use `$ vike build` instead of `$ vite build`
        //  - Remove the whole writeBundle() hook logic
        //  - make `$ vite build` only build the client-side
        return !(0, isViteCliCall_js_1.isViteCliCall)();
    }
    else {
        return disableAutoFullBuild;
    }
}
