// vike/dist/cjs/node/plugin/plugins/baseUrls.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.baseUrls = baseUrls;
const resolveBase_js_1 = require("../plugins/config/resolveBase.js");
const utils_js_1 = require("../utils.js");
const getConfigVike_js_1 = require("../../shared/getConfigVike.js");
function baseUrls(configVike) {
    let baseServer;
    let baseAssets;
    return {
        name: 'vike:baseUrls',
        enforce: 'post',
        async config(config) {
            const bases = (0, resolveBase_js_1.resolveBaseFromUserConfig)(config, configVike);
            baseServer = bases.baseServer;
            baseAssets = bases.baseAssets;
            // We cannot define these in configResolved() because Vite picks up the env variables before any configResolved() hook is called
            process.env.BASE_SERVER = baseServer;
            process.env.BASE_ASSETS = baseAssets;
            return {
                envPrefix: [
                    'VITE_', // Vite doesn't seem to merge in its default, see https://github.com/vikejs/vike/issues/554
                    'BASE_SERVER',
                    'BASE_ASSETS'
                ],
                base: baseAssets, // Make Vite inject baseAssets to imports e.g. `import logoUrl from './logo.svg.js'`
                _baseOriginal: config.base ?? '/__UNSET__' // Vite resolves `_baseOriginal: null` to `undefined`
            };
        },
        async configResolved(config) {
            const configVike = await (0, getConfigVike_js_1.getConfigVike)(config);
            // Ensure that the premature base URL resolving we did in config() isn't erroneous
            (0, utils_js_1.assert)(configVike.baseServer === baseServer);
            (0, utils_js_1.assert)(configVike.baseAssets === baseAssets);
            /* In dev, Vite seems buggy around setting vite.config.js#base to an absolute URL (e.g. http://localhost:8080/cdn/)
             *  - In dev, Vite removes the URL origin. (I.e. it resolves the user config `vite.config.js#base: 'http://localhost:8080/cdn/'` to resolved config `config.base === '/cdn/'`.)
             *  - Instead of having an internal Vike assertion fail, we let the user discover Vite's buggy behavior.
            assert(config.base === baseAssets)
            */
        }
    };
}
