// vike/dist/cjs/node/plugin/plugins/previewConfig.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.previewConfig = previewConfig;
const utils_js_1 = require("../utils.js");
const getConfigVike_js_1 = require("../../shared/getConfigVike.js");
const fs_1 = __importDefault(require("fs"));
const path_1 = __importDefault(require("path"));
const addSsrMiddleware_js_1 = require("../shared/addSsrMiddleware.js");
const picocolors_1 = __importDefault(require("@brillout/picocolors"));
function previewConfig() {
    let config;
    let configVike;
    return {
        name: 'vike:previewConfig',
        apply: 'serve',
        config(config) {
            return {
                appType: 'custom',
                build: {
                    outDir: (0, utils_js_1.resolveOutDir)(config)
                }
            };
        },
        async configResolved(config_) {
            config = config_;
            configVike = await (0, getConfigVike_js_1.getConfigVike)(config);
        },
        configurePreviewServer(server) {
            /* - Couldn't make `appType: 'mpa'` work as of npm:@brillout/vite@5.0.0-beta.14.0426910c
               - This ugly hack to set appType for preview won't be need once https://github.com/vitejs/vite/pull/14855 is merged.
            config.appType = 'mpa'
            */
            (0, utils_js_1.markEnvAsVitePreview)();
            return () => {
                assertDist();
                /* We don't use this condition (we wrongfully always use the SSR middleware) because of the regression introduced by https://github.com/vitejs/vite/pull/14756 which stops servering .html files when `appType: 'custom'`.
                if (!configVike.prerender || configVike.prerender.partial) {
                  addSsrMiddleware(server.middlewares, config, true)
                }
                /*/
                (0, addSsrMiddleware_js_1.addSsrMiddleware)(server.middlewares, config, true);
                //*/
                addStatic404Middleware(server.middlewares);
            };
        }
    };
    function assertDist() {
        let { outDirRoot, outDirClient, outDirServer } = (0, utils_js_1.getOutDirs)(config);
        [outDirRoot, outDirClient, outDirServer].forEach((outDirAny) => {
            (0, utils_js_1.assertUsage)(fs_1.default.existsSync(outDirAny), `Cannot run ${picocolors_1.default.cyan('$ vite preview')}: your app isn't built (the build directory ${picocolors_1.default.cyan(outDirAny)} is missing). Make sure to run ${picocolors_1.default.cyan('$ vite build')} before running ${picocolors_1.default.cyan('$ vite preview')}.`);
        });
    }
    function addStatic404Middleware(middlewares) {
        const { outDirClient } = (0, utils_js_1.getOutDirs)(config);
        middlewares.use(config.base, (_, res, next) => {
            const file = path_1.default.posix.join(outDirClient, './404.html');
            if (fs_1.default.existsSync(file)) {
                res.statusCode = 404;
                res.end(fs_1.default.readFileSync(file));
            }
            else {
                next();
            }
        });
    }
}
