// vike/dist/cjs/node/plugin/plugins/setGlobalContext.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.setGlobalContext = setGlobalContext;
const globalContext_js_1 = require("../../runtime/globalContext.js");
const utils_js_1 = require("../utils.js");
function setGlobalContext() {
    return {
        name: 'vike:setGlobalContext',
        enforce: 'pre',
        configureServer: {
            order: 'pre',
            handler(viteDevServer) {
                (0, globalContext_js_1.setGlobalContext_viteDevServer)(viteDevServer);
            }
        },
        config: {
            handler(_, env) {
                const isDev = (0, utils_js_1.isDev3)(env);
                (0, globalContext_js_1.setGlobalContext_isDev)(isDev);
            }
        },
        configResolved(config) {
            const { outDirRoot } = (0, utils_js_1.getOutDirs)(config);
            (0, utils_js_1.assertFilePathAbsoluteFilesystem)(outDirRoot); // Needed for loadImportBuild(outDir) of @brillout/vite-plugin-server-entry
            (0, globalContext_js_1.setGlobalContext_viteConfig)(config, outDirRoot);
        }
    };
}
