// vike/dist/esm/shared/page-configs/getConfigValueBuildTime.js
export { getConfigValueBuildTime };
import { assert } from '../utils.js';
import { getConfigValueTyped } from './getConfigValue.js';
import { assertIsNotProductionRuntime } from '../../utils/assertIsNotProductionRuntime.js';
assertIsNotProductionRuntime();
function getConfigValueBuildTime(pageConfig, configName, type) {
    const configValue = getConfigValue(pageConfig, configName);
    if (!configValue)
        return null;
    return getConfigValueTyped(configValue, configName, type);
}
function getConfigValue(pageConfig, configName) {
    const { configValueSources, configValuesComputed, configDefinitions } = pageConfig;
    const configValueComputed = configValuesComputed[configName];
    if (configValueComputed) {
        return {
            type: 'computed',
            value: configValueComputed.value,
            definedAtData: null
        };
    }
    const sources = configValueSources[configName];
    if (!sources)
        return null;
    const configDef = configDefinitions[configName];
    assert(configDef);
    if (!configDef.cumulative) {
        const configValueSource = sources[0];
        assert(configValueSource);
        assert(configValueSource.isOverriden === false);
        assert(sources.slice(1).every((s) => s.isOverriden === true));
        assert('value' in configValueSource);
        return {
            type: 'standard',
            value: configValueSource.value,
            definedAtData: getDefinedAtFile(configValueSource)
        };
    }
    else {
        const { value, definedAtData } = mergeCumulative(sources);
        assert(value.length === definedAtData.length);
        return {
            type: 'cumulative',
            value,
            definedAtData
        };
    }
}
function mergeCumulative(configValueSources) {
    const value = [];
    const definedAtData = [];
    configValueSources.forEach((configValueSource) => {
        assert(configValueSource.isOverriden === false);
        assert(configValueSource.configEnv.config === true);
        assert('value' in configValueSource);
        value.push(configValueSource.value);
        definedAtData.push(getDefinedAtFile(configValueSource));
    });
    return { value, definedAtData };
}
function getDefinedAtFile(configValueSource) {
    return {
        filePathToShowToUser: configValueSource.definedAtFilePath.filePathToShowToUser,
        fileExportPathToShowToUser: configValueSource.definedAtFilePath.fileExportPathToShowToUser
    };
}
